Tools/autotest/common.py

```diff
@@ -2339,137 +2339,147 @@ def delay_sim_time(self, seconds_to_wait):
+        self.wait_and_maintain(value_name="Heading", target=heading, achieved_getter=lambda: get_heading_wrapped(timeout), validator=lambda value2, target2: validator(value2, target2), accuracy=accuracy, timeout=timeout, **kwargs)
 
-    def wait_distance(self, distance, accuracy=5, timeout=30):
+    def wait_distance(self, distance, accuracy=2, timeout=30, **kwargs):
         """Wait for flight of a given distance."""
-        tstart = self.get_sim_time()
         start = self.mav.location()
-        last_distance_message = 0
-        while self.get_sim_time_cached() < tstart + timeout:
-            pos = self.mav.location()
-            delta = self.get_distance(start, pos)
-            if self.get_sim_time_cached() - last_distance_message >= 1:
-                self.progress("Distance=%.2f meters want=%.2f" %
-                              (delta, distance))
-                last_distance_message = self.get_sim_time_cached()
-            if math.fabs(delta - distance) <= accuracy:
-                self.progress("Attained distance %.02f meters OK" % delta)
-                return True
-            if delta > (distance + accuracy):
-                raise WaitDistanceTimeout("Failed distance - overshoot delta=%f dist=%f"
-                                          % (delta, distance))
-        raise WaitDistanceTimeout("Failed to attain distance %u" % distance)
 
-    def wait_distance_to_location(self, location, distance_min, distance_max, timeout=30):
-        last_distance_message = 0
-        tstart = self.get_sim_time()
-        while self.get_sim_time_cached() < tstart + timeout:
-            pos = self.mav.location()
-            delta = self.get_distance(location, pos)
-            if self.get_sim_time_cached() - last_distance_message >= 1:
-                self.progress("Distance=%.2f meters want <%.2f and >%.2f" %
-                              (delta, distance_min, distance_max))
-                last_distance_message = self.get_sim_time_cached()
-            if (delta >= distance_min and delta <= distance_max):
-                self.progress("Attained distance %.02f meters OK" % delta)
+        def get_distance():
+            return self.get_distance(start, self.mav.location())
+
+        def validator(value2, target2):
+            return (value2 - target2) >= accuracy
+
+        self.wait_and_maintain(value_name="Distance", target=distance, achieved_getter=lambda: get_distance(), validator=lambda value2, target2: validator(value2, target2), accuracy=accuracy, timeout=timeout, **kwargs)
+
+    def wait_distance_to_location(self, location, distance_min, distance_max, timeout=30, **kwargs):
+        """Wait for flight of a given distance."""
+        assert distance_min <= distance_max, "Distance min should be less than distance max."
+
+        def get_distance():
+            return self.get_distance(location, self.mav.location())
+
+        def validator(value2, target2=None):
+            if distance_min <= value2 <= distance_max:
```

Comment:
Do without the `if` here and just return the result?  Similarly in other validators/.